### PR TITLE
Database part1

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
PR 1/2
- add database components (PulseTable, ScheduleTemplate)

### CODE REVIEW

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 1]
```diff
@@ -27,32 +32,33 @@ class PulseTable:
...
-     (0,)   | d0      | x90p      | amp  | 0.03  | fail       | 2020.01.02
-    --------+---------+-----------+------+-------+------------+-------------
-     (0,1)  | u0      | cr90p     | amp  | 0.04  | pass       | 2020.01.01
-    ------------------------------------------------------------------------
-    ```
-
     You can search for the specific entry by filtering or you can directly generate
     keyword argument for the target pulse factory.
     """
+    TABLE_COLS = ['qubits', 'channel', 'inst_name', 'stretch', 'pulse_type',
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]:
```suggestion
    TABLE_COLS = ['qubits', 'channel', 'inst_name', 'stretch_factor', 'pulse_type',
```
`stretch` by itself is a little unclear.

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 2]
```diff
@@ -61,7 +67,9 @@ def get_dataframe(
         Args:
             qubits: Index of qubit(s) to search for.
             channel: Label of pulse channel to search for. Wildcards can be accepted.
-            gate_type: Name of gate to search for. Wildcards can be accepted.
+            inst_name: Name of gate to search for. Wildcards can be accepted.
+            pulse_type: Name of ParametricPulse generator that is used for pulse creation.
+            stretch: Stretch factor of the pulse.
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]:
```suggestion
            stretch_factor: Stretch factor of the pulse, typically used for error mitigation.
```
Can you change `stretch` to `stretch_factor` in the rest of the code?

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 3]
```diff
@@ -143,7 +159,7 @@ def get_generator_kwargs(
         # convert (amp, phase) pair into complex value
         if use_complex_amplitude:
             if 'amp' in format_dict and 'phase' in format_dict:
-                format_dict['amp'] *= np.exp(1j * format_dict.pop('phase'))
+                format_dict['amp'] *= np.exp(1j * format_dict['phase'])
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]: Will this not give errors if we do e.g. `Drag(**format_dict)` as `Drag` does not accept `phase`?

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 4]
```diff
@@ -98,7 +110,9 @@ def get_generator_kwargs(
         Args:
             qubits: Index of qubit(s) to search for.
             channel: Label of pulse channel to search for.
-            gate_type: Name of gate to search for.
+            inst_name: Name of gate to search for.
+            pulse_type: Name of ParametricPulse generator that is used for pulse creation.
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]: This is the name of the pulse envelope right? If so I don't think we should use the `generator` has it already has a strong meaning in the cal module.

Comment (eggerdj) [reply qiskit/ignis/experiments/calibration/instruction_data/database.py]: Can you fix this in the other comments too?

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 5]
```diff
@@ -235,19 +261,45 @@ def set_cal_data(
...
+        self._parameter_collection.at[data_index, 'validation'] = status
+
     def _find_data(
             self,
             qubits: Optional[Union[int, List[int]]] = None,
             channel: Optional[str] = None,
-            gate_type: Optional[str] = None,
+            inst_name: Optional[str] = None,
+            pulse_type: Optional[str] = None,
+            stretch: Optional[float] = None,
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]:
```suggestion
            stretch_factor: Optional[float] = None,
```
and update below

File: qiskit/ignis/experiments/calibration/instruction_data/database.py [hunk 6]
```diff
@@ -306,15 +366,130 @@ def _flatten(cls, entries: pd.DataFrame) -> Dict[str, List[types.CalValue]]:
...
+            init_dataframe = template_collection
+        else:
+            init_dataframe = pd.DataFrame(index=[], columns=ScheduleTemplate.TABLE_COLS)
 
-    # TODO: support parametrized schedule, eg Rx(theta) schedule
+        self._template_collection = init_dataframe
 
-    def __init__(self, templates: List[Dict[str, Union[str, pulse.Schedule]]]):
-        self._templates = templates
+    def get_dataframe(
```
Comment (eggerdj) [inline qiskit/ignis/experiments/calibration/instruction_data/database.py]: perhaps a name like `filter_data` would be more appropriate here?

**Review by eggerdj — CHANGES_REQUESTED** [review]
Looks good. Some of the naming can be improved a bit.

**Comment by nkanazawa1989** [comment]
Thanks!

**Review by eggerdj — APPROVED** [review]
LGTM

**Pull request merged**